File: javascript/two_pointers/11_Container_With_Most_Water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function(height) {
    let l = 0, r = height.length - 1;
    let res = 0;

    while (l < r) {
        area = Math.min(height[l], height[r]) * (r - l);
        res = Math.max(area, res);

        if (height[l] < height[r]) l++;
        else r--;
    }
    return res;
};

File: python/two_pointers/11_Container_With_Most_Water.py
'''
Brute force:
Time: O(n^2)
2 pointers:
Time: O(n)
'''

class Solution:
    # Brute force: time limit exceeded
    def maxArea(self, height: List[int]) -> int:
        res = 0

        for l in range(len(height)):  # left pointer
            for r in range(l+1, len(height)):
                area = (r - l) * min(height[l], height[r])
                res = max(area, res)

        return res

    # 2 pointers (l always < r)
    def maxArea(self, height: List[int]) -> int:
        l, r = 0, len(height) - 1
        res = 0

        while l < r:
            area = min(height[l], height[r]) * (r - l)
            res = max(area, res)

            if height[l] < height[r]:
                l += 1
            else: 
                r -= 1

        return res